### PR TITLE
[CLI] upgrade error formatting should use existing diag wrapper

### PR DESCRIPTION
**Pull request by jordanjennings-mysten**
## Description 

Reuse the work within `move-compiler/src/diagnostics/` instead of reimplementing code span reporting usage directly. 

## Test plan 

Snapshots

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui/src/lib.rs [hunk 1]
```diff
@@ -15,6 +15,7 @@ pub mod key_identity;
 pub mod keytool;
 pub mod shell;
 pub mod sui_commands;
+mod upgrade_compatibility;
```
Comment (jordanjennings-mysten) [inline crates/sui/src/lib.rs:18]: remove on merge.

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__declarations_missing.snap
Comment (amnn) [inline crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__declarations_missing.snap:1]: The change in error order here is kind of interesting -- is that expected? If the error order changes from run-to-run we may have issues with the snapshot test.

**Review by amnn — COMMENTED** [review]
Seems reasonable to me, but @tnowacki is the expert here, so I'll leave it with him to do the final review.

File: crates/sui/src/upgrade_compatibility.rs [hunk 2]
```diff
@@ -496,30 +497,19 @@ fn missing_definition_diag(
...
-            format!(
-                "Module '{module_name}' expected {declaration_kind} '{identifier_name}', but found none"
-            ),
-        )])
-        .with_notes(vec![format!(
-            "{declaration_kind}s are part of a module's public interface and cannot be removed or changed during an upgrade, add back the {declaration_kind} '{identifier_name}'."
-        )])
+        .definition_location;
+
+    Diagnostic::new(
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:507]: Is the `diag` macro exported from the compiler? You might find that easier to use (except for notes... maybe we should fix that)

**Review by tnowacki — CHANGES_REQUESTED** [review]
Requesting changes just to make it clear to change the `codes.rs` issue. Otherwise it should be good :) Thanks! Hopefully the switch wasn't too painful

File: crates/sui/src/upgrade_compatibility.rs (hunk 2 above)
Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:507]: I didn't come across this but I'll look around.

Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:507]: maybe a positional `notes = xyz`

File: external-crates/move/crates/move-compiler/src/diagnostics/codes.rs [hunk 3]
```diff
@@ -218,6 +218,7 @@ codes!(
         MissingSyntaxMethod: { msg: "no valid 'syntax' declaration found", severity: BlockingError },
         DuplicateAlias: { msg: "duplicate alias", severity: Warning },
         InvalidEnum: { msg: "invalid 'enum' declaration", severity: NonblockingError },
+        MissingPublic: { msg: "missing public declaration", severity: NonblockingError },
```
[the remark this answers is not in the document]
Comment (jordanjennings-mysten) [reply external-crates/move/crates/move-compiler/src/diagnostics/codes.rs:221]: I'm thinking `U` as the prefix? or maybe `UG` for upgrades?

[the remark this answers is not in the document]
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/diagnostics/codes.rs:221]: So far, we have spelled it out. So probably `"compatibility"` might be good

[the remark this answers is not in the document]
Comment (jordanjennings-mysten) [reply external-crates/move/crates/move-compiler/src/diagnostics/codes.rs:221]: that results in "error[**compatibility**E01001]: missing public declaration" in the error.

File: external-crates/move/crates/move-compiler/src/diagnostics/mod.rs [hunk 4]
```diff
@@ -747,6 +747,17 @@ macro_rules! diag {
             std::iter::empty::<String>(),
         )
     }};
+    ($code: expr, $primary: expr, note: $($note: expr),* $(,)?) => {{
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/diagnostics/mod.rs:750]: This isn't quite what I would have anticipated since it doesn't allow for secondary labels. Maybe this would be best suited for a different PR

File: external-crates/move/crates/move-compiler/src/diagnostics/codes.rs (hunk 3 above)
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/diagnostics/codes.rs:221]: We will still need to remove this

File: crates/sui/src/upgrade_compatibility.rs [hunk 5]
```diff
@@ -353,6 +359,56 @@ impl CompatibilityMode for CliCompatibilityMode {
...
+            }
+
+            // impl into diagnostic info
+            impl Into<DiagnosticInfo> for $cat {
+                fn into(self) -> DiagnosticInfo {
+                    match self {
+                        Self::ZeroPlaceholder =>
+                            panic!("do not use placeholder error code"),
+                        $(Self::$code => custom(
+                            "COMPATIBILITY",
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:392]:
```suggestion
                            "Compatibility ",
```

Comment (tnowacki) [reply crates/sui/src/upgrade_compatibility.rs:392]: And we might want to bind this to a constant somewhere. `const COMPATIBILITY_PREFIX: &str` or something similar.

File: crates/sui/src/upgrade_compatibility.rs [hunk 6]
```diff
@@ -353,6 +359,56 @@ impl CompatibilityMode for CliCompatibilityMode {
...
+        )*
+    };
+}
+
+// Used to generate diagnostics primary labels for upgrade compatibility errors.
+upgrade_codes!(
+    Declarations: [
+        PublicMissing: { msg: "missing public declaration" },
+    ],
+);
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:410]: I think in general this is fine, but the syntatic structure here was designed for a large amount of categories+codes with a lot of per-code tweaking, which isn't something we are doing here. 

We might want something a tad simpler like 
external-crates/move/crates/move-compiler/src/linters/mod.rs

But at the end of the day its up to you :)

Comment (jordanjennings-mysten) [reply crates/sui/src/upgrade_compatibility.rs:412]: It's probably overkill but I find it to be a relatively clean way to do it. Each category has its own set of codes which made the most sense to me.

File: crates/sui/Cargo.toml [hunk 7]
```diff
@@ -57,7 +57,7 @@ sui-cluster-test.workspace = true
 sui-execution = { path = "../../sui-execution" }
 sui-faucet.workspace = true
 sui-swarm-config.workspace = true
-sui-graphql-rpc = {workspace = true }
+sui-graphql-rpc = { workspace = true }
```
Comment (tnowacki) [inline crates/sui/Cargo.toml:60]: Either a revert or this might be nice
```suggestion
sui-graphql-rpc.workspace = true
```

File: crates/sui/src/upgrade_compatibility.rs [hunk 8]
```diff
@@ -353,6 +360,57 @@ impl CompatibilityMode for CliCompatibilityMode {
     }
 }
 
+const COMPATIBILITY_PREFIX: &str = "COMPATIBILITY";
```
Comment (tnowacki) [inline crates/sui/src/upgrade_compatibility.rs:363]:
```suggestion
const COMPATIBILITY_PREFIX: &str = "Compatibility ";
```

Comment (tnowacki) [reply crates/sui/src/upgrade_compatibility.rs:363]: I think `Compatibility E01001` reads a bit nicer than `COMPATIBILITYE01001`

File: crates/sui/src/unit_tests/fixtures/upgrade_errors/declarations_missing_v1/sources/func.move [hunk 9]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[allow(unused_field)]
```
Comment (tnowacki) [inline crates/sui/src/unit_tests/fixtures/upgrade_errors/declarations_missing_v1/sources/func.move:4]:
```suggestion
```

File: crates/sui/src/unit_tests/fixtures/upgrade_errors/declarations_missing_v2/sources/func.move [hunk 10]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[allow(unused_field)]
```
Comment (tnowacki) [inline crates/sui/src/unit_tests/fixtures/upgrade_errors/declarations_missing_v2/sources/func.move:4]:
```suggestion
```

File: crates/sui/src/unit_tests/fixtures/upgrade_errors/declarations_missing_v2/sources/struct.move [hunk 11]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[allow(unused_field)]
```
Comment (tnowacki) [inline crates/sui/src/unit_tests/fixtures/upgrade_errors/declarations_missing_v2/sources/struct.move:4]:
```suggestion
```

File: crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__declarations_missing.snap [hunk 12]
```diff
@@ -2,29 +2,35 @@
...
+  │                  ^^^^^ enum 'EnumToBeRemoved' is missing
   │
-  = structs are part of a module's public interface and cannot be removed or changed during an upgrade, add back the struct 'StructToBeRemoved'.
+  = enum is missing expected enum 'EnumToBeRemoved', but found none
+  = enums are part of a module's public interface and cannot be removed or changed during an upgrade
+  = add missing enum 'EnumToBeRemoved' back to the module 'enum_'.
 
-error: enum is missing
-  ┌─ /fixtures/upgrade_errors/declarations_missing_v2/sources/UpgradeErrors.move:7:18
+error[COMPATIBILITYE01001]: missing public declaration
```
Comment (tnowacki) [inline crates/sui/src/unit_tests/snapshots/sui__upgrade_compatibility__upgrade_compatibility_tests__declarations_missing.snap:15]: I think you will have to upgrade your snapshots here after the last change

**Review by tnowacki — APPROVED** [review]
Some small nits, but lgtm! Thanks for the changes :)
